File: php-checks/src/main/java/org/sonar/php/checks/AliasFunctionUsageCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2026 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.php.checks;

import java.util.Locale;
import java.util.Map;
import org.sonar.check.Rule;
import org.sonar.php.utils.collections.MapBuilder;
import org.sonar.plugins.php.api.tree.Tree;
import org.sonar.plugins.php.api.tree.declaration.NamespaceNameTree;
import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;

/**
 * @author Piotr Dawidiuk
 */
@Rule(key = AliasFunctionUsageCheck.KEY)
public class AliasFunctionUsageCheck extends PHPVisitorCheck {

  public static final String KEY = "S2050";

  public static final String MESSAGE = "Replace this use of \"%s\" with \"%s\".";

  private static final Map<String, String> ALIAS_FUNCTIONS = MapBuilder.<String, String>builder()
    .put("chop", "rtrim")
    .put("close", "closedir")
    .put("doubleval", "floatval")
    .put("fputs", "fwrite")
    .put("ini_alter", "ini_set")
    .put("is_double", "is_float")
    .put("is_integer", "is_int")
    .put("is_long", "is_int")
    .put("is_real", "is_float")
    .put("is_writeable", "is_writable")
    .put("join", "implode")
    .put("key_exists", "array_key_exists")
    .put("magic_quotes_runtime", "set_magic_quotes_runtime")
    .put("pos", "current")
    .put("show_source", "highlight_file")
    .put("sizeof", "count")
    .put("strchr", "strstr")
    .build();

  @Override
  public void visitFunctionCall(FunctionCallTree tree) {
    ExpressionTree callee = tree.callee();
    if (callee.is(Tree.Kind.NAMESPACE_NAME)) {
      String name = ((NamespaceNameTree) callee).qualifiedName();
      String replacementName = ALIAS_FUNCTIONS.get(name.toLowerCase(Locale.ROOT));
      if (replacementName != null) {
        context().newIssue(this, callee, String.format(MESSAGE, name, replacementName));
      }
    }

    super.visitFunctionCall(tree);
  }

}

File: php-checks/src/test/resources/checks/AliasFunctionUsageCheck.php
<?php

class Foo
{
  public function chop($param1) {
  }

  public function bar($param1, $param2)
  {
    $chop = 1;               // OK
    $this->chop($param1);    // OK

    chop($param1);           // Noncompliant {{Replace this use of "chop" with "rtrim".}}
//  ^^^^
    CHOP($param1);           // Noncompliant {{Replace this use of "CHOP" with "rtrim".}}
    rtrim($param1);          // OK

    close($param1);          // Noncompliant {{Replace this use of "close" with "closedir".}}
    closedir($param1);       // OK

    doubleval($param1);      // Noncompliant {{Replace this use of "doubleval" with "floatval".}}
    floatval($param1);       // OK

    fputs($param1);          // Noncompliant {{Replace this use of "fputs" with "fwrite".}}
    fwrite($param1);         // OK

    ini_alter($param1);      // Noncompliant {{Replace this use of "ini_alter" with "ini_set".}}
    ini_set($param1);        // OK

    is_double($param1);      // Noncompliant {{Replace this use of "is_double" with "is_float".}}
    is_float($param1);       // OK

    is_integer($param1);     // Noncompliant {{Replace this use of "is_integer" with "is_int".}}
    is_int($param1);         // OK

    is_long($param1);        // Noncompliant {{Replace this use of "is_long" with "is_int".}}
    is_int($param1);         // OK

    is_real($param1);        // Noncompliant {{Replace this use of "is_real" with "is_float".}}
    is_float($param1);       // OK

    is_writeable($param1);   // Noncompliant {{Replace this use of "is_writeable" with "is_writable".}}
    is_writable($param1);    // OK

    join($param1);           // Noncompliant {{Replace this use of "join" with "implode".}}
    implode($param1);        // OK

    key_exists($param1, $param2);        // Noncompliant {{Replace this use of "key_exists" with "array_key_exists".}}
    array_key_exists($param1, $param2);  // OK

    magic_quotes_runtime($param1);       // Noncompliant {{Replace this use of "magic_quotes_runtime" with "set_magic_quotes_runtime".}}
    set_magic_quotes_runtime($param1);   // OK

    pos($param1);            // Noncompliant {{Replace this use of "pos" with "current".}}
    current($param1);        // OK

    show_source($param1);    // Noncompliant {{Replace this use of "show_source" with "highlight_file".}}
    highlight_file($param1); // OK

    sizeof($param1);         // Noncompliant {{Replace this use of "sizeof" with "count".}}
    count($param1);          // OK

    strchr($param1);         // Noncompliant {{Replace this use of "strchr" with "strstr".}}
    strstr($param1);         // OK
 }

}
